### PR TITLE
[kube-dns] Fix CVEs

### PR DESCRIPTION
**Pull request by vitaliy-sn**
## Description

Fixed vulnerabilities: CVE-2022-1996, CVE-2022-27664, CVE-2022-41723, CVE-2023-39325, CVE-2022-32149, CVE-2021-33194, CVE-2021-38561, GHSA-m425-mq94-257g.

## Info

<details><summary>before</summary>

coredns
```shell
┌────────────────────────────────┬─────────────────────┬──────────┬────────────────────────────────────┬───────────────────────────────────┬─────────────────────────────────────────────────────────────┐
│            Library             │    Vulnerability    │ Severity │         Installed Version          │           Fixed Version           │                            Title                            │
├────────────────────────────────┼─────────────────────┼──────────┼────────────────────────────────────┼───────────────────────────────────┼─────────────────────────────────────────────────────────────┤
│ github.com/emicklei/go-restful │ CVE-2022-1996       │ CRITICAL │ v2.9.5+incompatible                │ 2.16.0                            │ Authorization Bypass Through User-Controlled Key            │
│                                │                     │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2022-1996                   │
├────────────────────────────────┼─────────────────────┼──────────┼────────────────────────────────────┼───────────────────────────────────┼─────────────────────────────────────────────────────────────┤
│ golang.org/x/net               │ CVE-2022-27664      │ HIGH     │ v0.0.0-20220520000938-2e3eb7b945c2 │ 0.0.0-20220906165146-f3363e06e74c │ handle server errors after sending GOAWAY                   │
│                                │                     │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2022-27664                  │
│                                ├─────────────────────┤          │                                    ├───────────────────────────────────┼─────────────────────────────────────────────────────────────┤
│                                │ CVE-2022-41723      │          │                                    │ 0.7.0                             │ avoid quadratic complexity in HPACK decoding                │
│                                │                     │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2022-41723                  │
│                                ├─────────────────────┤          │                                    ├───────────────────────────────────┼─────────────────────────────────────────────────────────────┤
│                                │ CVE-2023-39325      │          │                                    │ 0.17.0                            │ rapid stream resets can cause excessive work                │
│                                │                     │          │                                    │                                   │ (CVE-2023-44487)                                            │
│                                │                     │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2023-39325                  │
├────────────────────────────────┼─────────────────────┤          ├────────────────────────────────────┼───────────────────────────────────┼─────────────────────────────────────────────────────────────┤
│ golang.org/x/text              │ CVE-2022-32149      │          │ v0.3.7                             │ 0.3.8                             │ ParseAcceptLanguage takes a long time to parse complex tags │
│                                │                     │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2022-32149                  │
├────────────────────────────────┼─────────────────────┤          ├────────────────────────────────────┼───────────────────────────────────┼─────────────────────────────────────────────────────────────┤
│ google.golang.org/grpc         │ GHSA-m425-mq94-257g │          │ v1.46.2                            │ 1.56.3, 1.57.1, 1.58.3            │ gRPC-Go HTTP/2 Rapid Reset vulnerability                    │
│                                │                     │          │                                    │                                   │ https://github.com/advisories/GHSA-m425-mq94-257g           │
└────────────────────────────────┴─────────────────────┴──────────┴────────────────────────────────────┴───────────────────────────────────┴─────────────────────────────────────────────────────────────┘
```

sts-pods-hosts-appender-webhook
```shell
┌──────────────────────────┬────────────────┬──────────┬────────────────────────────────────┬───────────────────────────────────┬──────────────────────────────────────────────────────────────┐
│         Library          │ Vulnerability  │ Severity │         Installed Version          │           Fixed Version           │                            Title                             │
├──────────────────────────┼────────────────┼──────────┼────────────────────────────────────┼───────────────────────────────────┼──────────────────────────────────────────────────────────────┤
│ github.com/gogo/protobuf │ CVE-2021-3121  │ HIGH     │ v1.3.1                             │ 1.3.2                             │ plugin/unmarshal/unmarshal.go lacks certain index validation │
│                          │                │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2021-3121                    │
├──────────────────────────┼────────────────┤          ├────────────────────────────────────┼───────────────────────────────────┼──────────────────────────────────────────────────────────────┤
│ golang.org/x/net         │ CVE-2021-33194 │          │ v0.0.0-20201110031124-69a78807bb2b │ 0.0.0-20210520170846-37e1c6afe023 │ infinite loop in ParseFragment                               │
│                          │                │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2021-33194                   │
│                          ├────────────────┤          │                                    ├───────────────────────────────────┼──────────────────────────────────────────────────────────────┤
│                          │ CVE-2022-27664 │          │                                    │ 0.0.0-20220906165146-f3363e06e74c │ handle server errors after sending GOAWAY                    │
│                          │                │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2022-27664                   │
│                          ├────────────────┤          │                                    ├───────────────────────────────────┼──────────────────────────────────────────────────────────────┤
│                          │ CVE-2022-41723 │          │                                    │ 0.7.0                             │ avoid quadratic complexity in HPACK decoding                 │
│                          │                │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2022-41723                   │
│                          ├────────────────┤          │                                    ├───────────────────────────────────┼──────────────────────────────────────────────────────────────┤
│                          │ CVE-2023-39325 │          │                                    │ 0.17.0                            │ rapid stream resets can cause excessive work                 │
│                          │                │          │                                    │                                   │ (CVE-2023-44487)                                             │
│                          │                │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2023-39325                   │
├──────────────────────────┼────────────────┤          ├────────────────────────────────────┼───────────────────────────────────┼──────────────────────────────────────────────────────────────┤
│ golang.org/x/text        │ CVE-2021-38561 │          │ v0.3.4                             │ 0.3.7                             │ out-of-bounds read in golang.org/x/text/language leads to    │
│                          │                │          │                                    │                                   │ DoS                                                          │
│                          │                │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2021-38561                   │
│                          ├────────────────┤          │                                    ├───────────────────────────────────┼──────────────────────────────────────────────────────────────┤
│                          │ CVE-2022-32149 │          │                                    │ 0.3.8                             │ ParseAcceptLanguage takes a long time to parse complex tags  │
│                          │                │          │                                    │                                   │ https://avd.aquasec.com/nvd/cve-2022-32149                   │
└──────────────────────────┴────────────────┴──────────┴────────────────────────────────────┴───────────────────────────────────┴──────────────────────────────────────────────────────────────┘
```

</details>

<details><summary>after</summary>

```shell
+ trivy i --severity=CRITICAL,HIGH dev-registry.deckhouse.io/sys/deckhouse-oss@sha256:1d493b463f63eb8a8ab8c94685669306341450bc959f2969f2cad5d7bac9785f
2023-11-03T09:59:54.400+0300    INFO    Vulnerability scanning is enabled
2023-11-03T09:59:54.400+0300    INFO    Secret scanning is enabled
2023-11-03T09:59:54.400+0300    INFO    If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-11-03T09:59:54.400+0300    INFO    Please see also https://aquasecurity.github.io/trivy/v0.38/docs/secret/scanning/#recommendation for faster secret detection
2023-11-03T09:59:55.060+0300    INFO    Number of language-specific files: 1
2023-11-03T09:59:55.060+0300    INFO    Detecting gobinary vulnerabilities...
```

</details>

<details><summary>annotation check</summary>

```shell
root@dev-master-0:~# kubectl get svc test -o yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
  creationTimestamp: "2023-11-10T05:08:19Z"
  name: test
  namespace: default
  resourceVersion: "27774272"
  uid: 0dbb6655-7ff6-4d26-bf44-51c51c0065c6
spec:
  clusterIP: None
  clusterIPs:
  - None
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: https
    port: 443
    protocol: TCP
    targetPort: 8443
  selector:
    app: test
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
root@dev-master-0:~# kubectl get ep 
NAME         ENDPOINTS            AGE
kubernetes   192.168.4.180:6443   35d
test                              38m
root@dev-master-0:~# kubectl exec -ti test-7848b7ddfc-m8tb9 -c test bash
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
root@test-7848b7ddfc-m8tb9:~# host -t a test
test.default.svc.cluster.local has address 10.111.4.149
root@test-7848b7ddfc-m8tb9:~# 
exit
root@dev-master-0:~# kubectl get po -o wide
NAME                    READY   STATUS    RESTARTS   AGE   IP             NODE                              NOMINATED NODE   READINESS GATES
test-7848b7ddfc-m8tb9   1/2     Running   0          39m   10.111.4.149   dev-worker-58bd4cc2-56b6f-fjcrx   <none>           <none>
```

</details>

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: kube-dns
type: fix
summary: Fixed vulnerabilities: CVE-2022-1996, CVE-2022-27664, CVE-2022-41723, CVE-2023-39325, CVE-2022-32149, CVE-2021-33194, CVE-2021-38561.
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/static'...

**Comment by vitaliy-sn** [comment]
/e2e/abort static;Static;containerd;1.25 6794711247 6794711247-1-con-1-25 /sys/deckhouse-oss/install:pr6397 astra@192.168.199.206

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/aws'...

**Comment by vitaliy-sn** [comment]
/e2e/abort aws;WithoutNAT;containerd;1.25 6795017298 6795017298-1-con-1-25 /sys/deckhouse-oss/install:pr6397

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/static'...

**Comment by RomanenkoDenys** [comment]
Please check annotation work

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/static'...

File: modules/042-kube-dns/images/coredns/patches/support-alpha-tolerate-unready-endpoints-annotation.patch [hunk 1]
```diff
@@ -1,145 +1,42 @@
-diff --git a/plugin/kubernetes/handler_test.go b/plugin/kubernetes/handler_test.go
```
Comment (name212) [inline modules/042-kube-dns/images/coredns/patches/support-alpha-tolerate-unready-endpoints-annotation.patch:1]: what is it? I talk about IDEA additional info

**Review by name212 — COMMENTED** [review]
one thing

**Review by name212 — APPROVED** [review]

**Pull request merged**